### PR TITLE
Upgrade to React 18

### PR DESCRIPTION
**Pull request by koesie10**
This upgrades the webviews to use React 18. There should be no user-facing changes.

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

**Review by charisk — APPROVED** [review]

**Comment by koesie10** [comment]
Thanks for the review @charisk! I've had to make some other changes to make it fully work with React 18. The results view wasn't working due to how React is now batching state updates, so there were quite some changes in how the results view now receives some messages.

@aeisenberg Would you be able to review those changes as well? It's mostly about https://github.com/github/vscode-codeql/pull/2167/commits/36eb2cd2eaeb1e8312e4aea9ed171e793e0b8b4d. I don't believe there are any major behaviour changes, but we're not using `shouldKeepOldResultsWhileRendering` anymore since we're loading the results in the same `setState` call as where that gets set.

File: extensions/ql-vscode/src/view/webview.tsx [hunk 1]
```diff
@@ -25,11 +26,11 @@ const render = () => {
...
-  ReactDOM_render(
-    view.component,
-    document.getElementById("root"),
-    // Post a message to the extension when fully loaded.
-    () => vscode.postMessage({ t: "viewLoaded", viewName }),
+  createRoot(element).render(
+    // Post a message to the extension when fully loaded. See https://github.com/reactwg/react-18/discussions/5 ("What about the render callback?")
+    <div ref={() => vscode.postMessage({ t: "viewLoaded", viewName })}>
+      {view.component}
+    </div>,
```
Comment (elenatanasoiu) [inline extensions/ql-vscode/src/view/webview.tsx:33]: Shouldn't we be changing `view.component` to accept callbacks and then do:

```
const rootElement = createRoot(element);
rootElement.render(<view.component callback={() => vscode.postMessage({ t: "viewLoaded", viewName })} >)
```

I think at least having the `rootElement` in it's own variable makes it slight easier to parse.

File: extensions/ql-vscode/src/view/webview.tsx [hunk 2]
```diff
@@ -25,11 +26,11 @@ const render = () => {
   // eslint-disable-next-line @typescript-eslint/no-var-requires
   const view: WebviewDefinition = require(`./${viewName}/index.tsx`).default;
 
-  ReactDOM_render(
-    view.component,
-    document.getElementById("root"),
-    // Post a message to the extension when fully loaded.
-    () => vscode.postMessage({ t: "viewLoaded", viewName }),
+  createRoot(element).render(
+    // Post a message to the extension when fully loaded. See https://github.com/reactwg/react-18/discussions/5 ("What about the render callback?")
```
Comment (elenatanasoiu) [inline extensions/ql-vscode/src/view/webview.tsx:30]: I'm not sure we need the comment to stick around. We could have it in the commit message for people to find if they want to know why we made the change.

File: extensions/ql-vscode/src/view/results/results.tsx [hunk 3]
```diff
@@ -164,76 +162,52 @@ export class ResultsApp extends React.Component<
 
   private updateStateWithNewResultsInfo(resultsInfo: ResultsInfo): void {
     this.setState((prevState) => {
-      const stateWithDisplayedResults = (
```
Comment (elenatanasoiu) [inline extensions/ql-vscode/src/view/results/results.tsx:167]: Could we write tests for the behaviour before we made this change and then see if the same behaviour happens afterwards? That would mean we wouldn't need to guess if we've completely got it right.

**Comment by elenatanasoiu** [comment]
Do we get any extra warnings now that we've upgraded to React 18? I'd be surprised if we have a clean upgrade.

File: extensions/ql-vscode/src/view/webview.tsx (hunk 1 above)
Comment (koesie10) [reply extensions/ql-vscode/src/view/webview.tsx:33]: I think localising this behaviour to the `webview.tsx` file makes the rest of the components much clearer. I don't think the other components should be concerned with how the `viewLoaded` message gets send and adding an extra `div` shouldn't change anything about the layout.

I will extract the `createRoot` result into a separate variable, this also seems to be shown in [the React documentation](https://beta.reactjs.org/reference/react-dom/client/createRoot#createroot) with a name of `root`. I think `rootElement` could be confusing because you could confuse it with an HTML element. The type name of the return type of `createRoot` is `Root`, so calling it `root` like in the documentation seems like it would make sense.

File: extensions/ql-vscode/src/view/webview.tsx (hunk 2 above)
Comment (koesie10) [reply extensions/ql-vscode/src/view/webview.tsx:30]: I left the comment here because this is quite a strange use of the `ref` property. `ref` is usually used to get a reference to the underlying HTML element, and we're using it in a completely different way. I can definitely remove it and we can rely on the Git history.

**Comment by koesie10** [comment]
> Do we get any extra warnings now that we've upgraded to React 18? I'd be surprised if we have a clean upgrade.

We don't have any extra warnings since we're actually not using that many React features. What kind of warnings would you have expected to see?

File: extensions/ql-vscode/src/view/webview.tsx (hunk 2 above)
Comment (elenatanasoiu) [reply extensions/ql-vscode/src/view/webview.tsx]: That's fair, I didn't clock from the link that `ref` is required now. Should we extract this div into an `App` component to more closely match the docs example?

**Comment by elenatanasoiu** [comment]
I was expecting to see warnings when we run `npm run build` / `npm run watch`, during CI or when running tests locally. But if those are clean, that's great. I also see you''re enabling strict mode in a different PR so maybe that will show more warnings in the future.

File: extensions/ql-vscode/src/view/webview.tsx (hunk 1 above)
Comment (elenatanasoiu) [reply extensions/ql-vscode/src/view/webview.tsx:33]: Sounds good, and yeah adding it to view.component doesn't quite make sense. Perhaps we can turn the extra `div` into an `App`-like component, like they have in the docs?

File: extensions/ql-vscode/src/view/results/results.tsx (hunk 3 above)
Comment (elenatanasoiu) [reply extensions/ql-vscode/src/view/results/results.tsx:167]: Is it hard because we would have to check visual components or because it's a timing issue?

Comment (koesie10) [reply extensions/ql-vscode/src/view/results/results.tsx:167]: It's a timing issue, there's no real way to check the state in between the two `setState` calls, so we'd never get into the state that we'd want to test. We'd want to check what the view looks like in between [these two calls](https://github.com/github/vscode-codeql/blob/64d97aaf7e47f1a012d4564159193b767ff14eb0/extensions/ql-vscode/src/view/results/results.tsx#L93-L107) but they are called sequentially without any waiting, so there's no way to actually test that state.

Comment (elenatanasoiu) [reply extensions/ql-vscode/src/view/results/results.tsx:167]: I'm gonna defer to @aeisenberg on this one. I would ask if we can get rid of `shouldKeepOldResultsWhileRendering` entirely if we're combining calls.

File: extensions/ql-vscode/src/view/jest.setup.ts [hunk 4]
```diff
@@ -15,6 +15,11 @@ Object.defineProperty(window, "matchMedia", {
   })),
 });
 
+// Used by Primer React
+(window as any).CSS = {
+  supports: jest.fn().mockResolvedValue(false),
```
Comment (elenatanasoiu) [inline extensions/ql-vscode/src/view/jest.setup.ts:20]: What are we mocking here exactly? If this always returns false, are we saying we won't support Primer CSS? My primer-fu is failing me on this one. 

I can't see primer/css installed in package.json so I'm assuming it's a dependency of `primer/react` or some other `@primer/*` package. 

I'm also aware we want to move away from Primer entirely and it looks like only one component uses it, through `ThemeProvider`. I'm wondering if we could make a copy of the `ThemeProvider` component and delete primer entirely. One for later probably.

Comment (koesie10) [reply extensions/ql-vscode/src/view/jest.setup.ts:20]: We're mocking `CSS.supports` here to make these calls work: https://github.com/primer/react/blob/f0e0ec77f5d91c73ce22d10bc8d55a1a5ab2a9e6/src/PageLayout/useStickyPaneHeight.ts#L106-L113. We're essentially saying that we don't support any modern CSS features, see [the MDN documentation for `CSS.supports`](https://developer.mozilla.org/en-US/docs/Web/API/CSS/supports). This won't change any of the behaviour, it just makes it look like we're using an older browser (which `jest-dom` essentially is since it doesn't support CSS at all).

The `ThemeProvider` code doesn't look like it uses many other files from `@primer/react`: https://github.com/primer/react/blob/f0e0ec77f5d91c73ce22d10bc8d55a1a5ab2a9e6/src/ThemeProvider.tsx. However, we'd also need to copy over the `Overlay` component. This seems to have a lot of dependencies on other files in `@primer/react`, so we'd also need to copy over those: https://github.com/primer/react/blob/f0e0ec77f5d91c73ce22d10bc8d55a1a5ab2a9e6/src/Overlay.tsx

@charisk is working on a new CodePaths view that doesn't use the `Overlay` component, so we can get rid of Primer once that work is done. That would also allow us to get rid of this mock.

File: extensions/ql-vscode/src/view/results/results.tsx (hunk 3 above)
Comment (elenatanasoiu) [reply extensions/ql-vscode/src/view/results/results.tsx:167]: Looks like Andrew is on holiday so you might have to wait a bit with this one.

Comment (charisk) [reply extensions/ql-vscode/src/view/results/results.tsx:167]: Just wanted to say that I wasn't expecting us to do a complete re-write of this view soon as the changes we'd like to make are mostly cosmetic. But I can be convinced otherwise.

**Comment by charisk** [comment]
> Thanks for the review @charisk! I've had to make some other changes to make it fully work with React 18. The results view wasn't working due to how React is now batching state updates, so there were quite some changes in how the results view now receives some messages.

Thanks for the poke! The changes look okay to me, but as you say the logic is quite hard to follow. I've also tested them a bit manually and things seem ok to me.

**Comment by elenatanasoiu** [comment]
Thanks for testing @charisk! I wasn't sure how to go about checking this. If we know what to expect in the behaviour change, I don't want to hold this up.

**Review by elenatanasoiu — APPROVED** [review]

File: extensions/ql-vscode/src/view/results/results.tsx [hunk 5]
```diff
@@ -164,76 +162,52 @@ export class ResultsApp extends React.Component<
...
+          database: resultsInfo.database,
+          sortStates: this.getSortStates(resultsInfo),
+        };
+      } catch (e) {
+        let errorMessage: string;
+        if (e instanceof Error) {
+          errorMessage = e.message;
+        } else {
+          errorMessage = "Unknown error";
+        }
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/view/results/results.tsx:206]:
```suggestion
        const errorMessage = getErrorMessage(e);
```

File: extensions/ql-vscode/src/view/results/results.tsx [hunk 6]
```diff
@@ -164,76 +162,52 @@ export class ResultsApp extends React.Component<
...
+          return {
+            displayedResults: {
+              resultsInfo: null,
+              results: null,
+              errorMessage: "No results to display",
+            },
+            isExpectingResultsUpdate: prevState.isExpectingResultsUpdate,
+            nextResultsInfo: resultsInfo,
+          };
+        }
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/view/results/results.tsx:188]: Are you missing a case here? I don't see line 191 from the original handled.

```
return stateWithDisplayedResults(prevState.displayedResults);
```

File: extensions/ql-vscode/src/view/results/results.tsx (hunk 3 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/view/results/results.tsx:167]: This method is overly complex and a simplification would be really nice. To be safe when we are doing a refactoring like this, it would be nice to add some pure regression unit tests to ensure that all the different cases are handled identically before and after.

Although nice, it's not necessary to spend the time to create more complex tests that check the view between the two `setState` calls. All that I think is necessary is that we ensure the method behaves the same before and after the refactoring. It's a complicated method, but at least the complexity is contained.

Comment (koesie10) [reply extensions/ql-vscode/src/view/results/results.tsx:167]: I've created a really simple regression test in #2255. This does not test the loading state, but checks that we can at least render a SARIF file.

File: extensions/ql-vscode/src/view/results/results.tsx (hunk 6 above)
Comment (koesie10) [reply extensions/ql-vscode/src/view/results/results.tsx:188]: In this case, the missing case is correct. In the previous code, after the `setState` returned with `stateWithDisplayedResults(prevState.displayedResults)`, `this.loadResults()` would be called immediately. The only reason for storing the `nextResultsInfo` in the state is so that `loadResults` can use it. However, with concurrent updates the state seen in `loadResults` would not yet be the state set in this `setState` call. Therefore, I've combined this `setState` and the `loadResults` method. Therefore, this case won't happen anymore.

Thanks for the suggestion for the simplification. Since we don't need the third case, I couldn't use it as-is. However, I have split the flows differently so there's only 1 level of nesting rather than 2.

**Review by aeisenberg — COMMENTED** [review]
This looks good now, but can you rebase on https://github.com/github/vscode-codeql/pull/2255 so we can be sure that the behaviour remains consistent?
